esphome/components/dsmr/parser.h

```diff
@@ -0,0 +1,417 @@
+#include "util.h"
+
+namespace dsmr {
+
+/**
+ * ParsedData is a template for the result of parsing a Dsmr P1 message.
+ * You pass the fields you want to add to it as template arguments.
+ *
+ * This template will then generate a class that extends all the fields
+ * passed (the fields really are classes themselves). Since each field
+ * class has a single member variable, with the same name as the field
+ * class, all of these fields will be available on the generated class.
+ *
+ * In other words, if I have:
+ *
+ * using MyData = ParsedData<
+ *  identification,
+ *  equipment_id
+ * >;
+ *
+ * MyData data;
+ *
+ * then I can refer to the fields like data.identification and
+ * data.equipment_id normally.
+ *
+ * Furthermore, this class offers some helper methods that can be used
+ * to loop over all the fields inside it.
+ */
+template<typename... Ts> struct ParsedData;
+
+/**
+ * Base case: No fields present.
+ */
+template<> struct ParsedData<> {
+  ParseResult<void> __attribute__((__always_inline__))
+  parse_line_inlined(const ObisId & /* id */, const char *str, const char * /* end */) {
+    // Parsing succeeded, but found no matching handler (so return
+    // set the next pointer to show nothing was parsed).
+    return ParseResult<void>().until(str);
+  }
+
+  template<typename F> void __attribute__((__always_inline__)) applyEach_inlined(F && /* f */) {
+    // Nothing to do
+  }
+
+  bool all_present_inlined() { return true; }
+};
+
+// Do not use F() for multiply-used strings (including strings used from
+// multiple template instantiations), that would result in multiple
+// instances of the string in the binary
+static constexpr char DUPLICATE_FIELD[] DSMR_PROGMEM = "Duplicate field";
+
+/**
+ * General case: At least one typename is passed.
+ */
+template<typename T, typename... Ts> struct ParsedData<T, Ts...> : public T, ParsedData<Ts...> {
+  /**
+   * This method is used by the parser to parse a single line. The
+   * OBIS id of the line is passed, and this method recursively finds a
+   * field with a matching id. If any, it calls it's parse method, which
+   * parses the value and stores it in the field.
+   */
+  ParseResult<void> parse_line(const ObisId &id, const char *str, const char *end) {
+    return parse_line_inlined(id, str, end);
+  }
+
+  /**
+   * always_inline version of parse_line. This is a separate method, to
+   * allow recursively inlining all calls, but still have a non-inlined
+   * top-level parse_line method.
+   */
+  ParseResult<void> __attribute__((__always_inline__))
+  parse_line_inlined(const ObisId &id, const char *str, const char *end) {
+    if (id == T::id) {
+      if (T::present())
+        return ParseResult<void>().fail((const __FlashStringHelper *) DUPLICATE_FIELD, str);
+      T::present() = true;
+      return T::parse(str, end);
+    }
+    return ParsedData<Ts...>::parse_line_inlined(id, str, end);
+  }
+
+  template<typename F> void applyEach(F &&f) { applyEach_inlined(f); }
```

Comment:
```suggestion
  template<typename F> void apply_each(F &&f) { apply_each_inlined(f); }
```

esphome naming style.